### PR TITLE
Add SpaceNet5

### PR DESCRIPTION
**Pull request by ashnair1**
- Added SpaceNet 5
- Added tests and test data for SpaceNet 5
- `_load_mask` and `plot` overrided to support multi channel speed masks.

Sample Plots:

Binary Mask (With speed_mask=`False`)
binary_mask

Speed Mask (With speed_mask=`True`)
speed_mask

By default, every 10 mph constitutes a class hence 7 classes (not including background).

### CODE REVIEW

File: tests/datasets/test_spacenet.py [hunk 1]
```diff
@@ -223,6 +223,117 @@ def test_plot(self, dataset: SpaceNet4) -> None:
         plt.close()
 
 
+class TestSpaceNet5:
+    @pytest.fixture(params=["PAN", "MS", "PS-MS", "PS-RGB"])
```
Comment (adamjstewart) [inline tests/datasets/test_spacenet.py:227]: Let's combine these two fixtures into a single fixture. Your `request.param` can be a list of attributes, so you can use `zip` or `itertools.product` or something like that to add more combinations.

File: torchgeo/datasets/spacenet.py [hunk 2]
```diff
@@ -737,6 +738,242 @@ def _load_files(self, root: str) -> List[Dict[str, str]]:
...
+        max_speed_bin = 65
+        speed_arr_bin = np.arange(min_speed_bin, max_speed_bin + 1)
+        bin_size_mph = 10.0
+        speed_cls_arr = np.array(
+            [int(math.ceil(s / bin_size_mph)) for s in speed_arr_bin]
+        )
+
+        try:
+            with fiona.open(path) as src:
+                vector_crs = CRS(src.crs)
```
Comment (adamjstewart) [inline torchgeo/datasets/spacenet.py:888]: Isn't `src.crs` already a `CRS` object?

File: torchgeo/datasets/spacenet.py [hunk 3]
```diff
@@ -737,6 +738,242 @@ def _load_files(self, root: str) -> List[Dict[str, str]]:
...
+                            vector_crs.to_string(),
+                            raster_crs.to_string(),
+                            feature["geometry"],
+                        )
+                    else:
+                        geom = feature["geometry"]
+
+                    if self.speed_mask:
+                        val = speed_cls_arr[
+                            int(feature["properties"]["inferred_speed_mph"]) - 1
```
Comment (adamjstewart) [inline torchgeo/datasets/spacenet.py:903]: Why -1?

File: torchgeo/datasets/spacenet.py [hunk 4]
```diff
@@ -737,6 +738,242 @@ def _load_files(self, root: str) -> List[Dict[str, str]]:
...
+            path: path to the label
+            tfm: transform of corresponding image
+            shape: shape of corresponding image
+
+        Returns:
+            Tensor: label tensor
+        """
+        min_speed_bin = 1
+        max_speed_bin = 65
+        speed_arr_bin = np.arange(min_speed_bin, max_speed_bin + 1)
```
Comment (adamjstewart) [inline torchgeo/datasets/spacenet.py:880]: Why +1?

File: torchgeo/datasets/spacenet.py [hunk 5]
```diff
@@ -737,6 +738,242 @@ def _load_files(self, root: str) -> List[Dict[str, str]]:
...
+        bin_size_mph = 10.0
+        speed_cls_arr = np.array(
+            [int(math.ceil(s / bin_size_mph)) for s in speed_arr_bin]
```
Comment (adamjstewart) [inline torchgeo/datasets/spacenet.py:883]: Why are we binning these? Wouldn't it be better to leave them as floats and use it in a regression problem?

File: torchgeo/datasets/spacenet.py (hunk 3 above)
Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:903]: `speed_cls_arr` is an array of length 65 with a class corresponding to each index/speed mph. Essentially its a mapping between speed mph (index) and the class. -1 is to account for index starting at 0.

File: torchgeo/datasets/spacenet.py (hunk 4 above)
Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:880]: `np.arange(min, max)` produces an array ranging from `min` to `max-1`. +1 is to include the max

File: torchgeo/datasets/spacenet.py (hunk 2 above)
Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:888]: With fiona, it's epsg dict i.e. `{'init': 'epsg:4326'}`

File: torchgeo/datasets/spacenet.py (hunk 5 above)
Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:883]: Binning is done to create a multi channel speed mask which would enable training a segmentation model. This was the strategy adopted by the SpaceNet authors (cresi) and the winning solutions hence why I opted for this,

File: tests/datasets/test_spacenet.py (hunk 1 above)
Comment (ashnair1) [reply tests/datasets/test_spacenet.py]: Done

File: tests/datasets/test_spacenet.py [hunk 6]
```diff
@@ -223,6 +224,77 @@ def test_plot(self, dataset: SpaceNet4) -> None:
         plt.close()
 
 
+class TestSpaceNet5:
+    @pytest.fixture(
+        params=list(itertools.product(["PAN", "MS", "PS-MS", "PS-RGB"], [False, True]))
```
Comment (adamjstewart) [inline tests/datasets/test_spacenet.py:229]:
```suggestion
        params=itertools.product(["PAN", "MS", "PS-MS", "PS-RGB"], [False, True])
```

**Review by adamjstewart — DISMISSED** [review]

**Review by adamjstewart — DISMISSED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by calebrob6** [comment]
Good to merge @ashnair1?

**Comment by ashnair1** [comment]
> Good to merge @ashnair1?

Good to go 👍

**Pull request merged**